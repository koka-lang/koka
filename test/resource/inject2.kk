// Test inject
fun injectst( action : () -> e a ) : <st<h>|e> a {
  // we can inject using an explicit `some`, or just underscore, or leave out the parameters
  inject<alloc>{inject<write<_h>>{inject<some<h>read<h>>(action)}}
}

fun xforeach-indexed( xs : list<a>, action : (int,a) -> e () ) : e ()
{
  var i := 0
  xs.foreach fun(x) {
    val j = i;
    injectst{action(j,x)}
    i := i+1
  }
}

fun main() {
  var acc := 0
  [0,41].xforeach-indexed fun(idx,x) {
    acc := acc + idx + x
  }
  println(acc)
}