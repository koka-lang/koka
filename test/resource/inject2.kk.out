42
injectst : forall<a,e,h> (action : () -> e a) -> <st<h>|e> a
main : () -> console ()
xforeach-indexed: forall<a,e> (xs : list<a>, action : (int, a) -> e ()) -> e ()